Create organization on user signup

jeremyzitomer-usds opened the pull request:
## Related Issue or Background Info

Fixes #1325

- https://app.zenhub.com/workspaces/simplereport-growth--engagement-605b43a6ceb92c000f86279a/issues/cdcgov/prime-simplereport/1325

## Changes Proposed

- Create organization, facility, and (inactive) Okta user on account sign up
- In demo environments, create Okta entities for all orgs/facilities in DB at startup
- Created magic user for auditing purposes when someone creates organization/user via account request page
- Created `CurrentAccountRequestContextHolder` to flag account requests for auditing purposes
- Changed security restrictions for create org/create user
- Added descriptive errors when user submits unregistered devices in new org account request
- Adjusted some tests

## Additional information
- This MUST be deployed in tandem with https://github.com/CDCgov/prime-simplereport-site/pull/127

## Checklist for Author and Reviewer
- [x] Run static site locally under this branch: https://github.com/CDCgov/prime-simplereport-site/pull/127
- [x] Submit new org using static site's sign-up form
- [x] Run core app locally as a Super User using the branch jeremyzitomer-usds/create-org-signup
- [x] Verify that the org created from the sign up form appears in list of unverified orgs on admin page of core app
- [x] As a bonus, if you want: copy the external ID of the new org you just created, mark that new unverified org as verified, and then go change your application-local.yaml to make the default user the new user/org you just created using the newly created username and org external ID and role=ADMIN, relaunch the app and verify that your app loads correctly while loaded from the view of the newly created user

### UI
- [x] Any changes to the UI/UX have been approved by design 
- [x] Any new or edited error messages have been approved by design 

### Testing
- [x] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [x] Database changes are submitted as a separate PR
- [x] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [x] Any dependencies introduced have been vetted and discussed

## Cloud
- [x] DevOps team has been notified if PR requires ops support

emmastephenson commented on backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java:
```diff
@@ -56,11 +88,90 @@ public void submitAccountRequest(@Valid @RequestBody AccountRequest body) throws
+                HashMap::new, 
+                (m,e)->m.put(
+                    e.getKey(), 
+                    e.getValue() == null 
```
Should we also check if the string is empty here?

emmastephenson commented on backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java:
```diff
@@ -56,11 +88,90 @@ public void submitAccountRequest(@Valid @RequestBody AccountRequest body) throws
+
+    List<String> testingDevicesSubmitted = 
+        new ArrayList<>(Arrays.asList(reqVars.get("testingDevices").split(", ")));
+    testingDevicesSubmitted.removeIf(d -> d.toLowerCase().startsWith("other"));
```
Is it possible to store this "other" field somewhere? Right now it seems like we're dropping the value, so that when the user goes in to see their org the device won't appear.
Also, if the user has only selected "other" will the graphQL throw an error? I assume organizations need at least one device.

jeremyzitomer-usds replied:
a) actually by design, yes, we're not storing the "other" device. it gets passed onto the account request email (in account-request.html) but from a SimpleReport app perspective, there's not much we can do to accommodate those "other" devices anyway at the moment. only the "pre-approved" devices get stored.
b) the sign-up form forces the user to choose a default device from the list of pre-approved devices, and the app enforces that the default device has to be in the user's list of devices, so effectively the answer is yes, the app would throw an error. but it'll come back to the user as a non-descript 5xx error since it's not wired through graphQL like our core app APIs :-\

emmastephenson replied:
Ahhh ok, I didn't realize that the signup form forces users to choose a default device from the list of acceptable devices, that makes much more sense. Thanks for the detailed explanation!

ClydeDroid reviewed (approved):
This all looks pretty good to me, though interested in Emma's notes as well. Great work!

jeremyzitomer-usds commented:
@emmastephenson @nickclyde @bmaden-usds this is now ready for review!

jeremyzitomer-usds commented:
had to make a number of changes to this PR due to bugs i found on `test` over the weekend, so i need to ask @emmastephenson @bmaden-usds @nickclyde to take one more look at this in the early part of today, if you can. there are some security-related changes too. i'd like to deploy by EOD.
new changes:
- In demo environments, create Okta entities for all existing orgs/facilities in DB at startup
- Created magic user for auditing purposes when someone creates organization/user via account request page (borrowed straight from @nullflux 's other magic_user PR)
- Created `CurrentAccountRequestContextHolder` to flag account requests for auditing purposes
- Changed security restrictions for create org/create user -- described in Slack. the PR checklist says we have to run this change by security engineers -- is that @jeskew-gov ?
- Added descriptive errors when user submits unregistered devices in new org account request

nickclyde reviewed (approved):
All the changes look good to me :+1:

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java:
```diff
@@ -47,20 +82,119 @@ public void submitWaitlistRequest(@Valid @RequestBody WaitlistRequest body) thro
+            "%s-%s-%s",
+            reqVars.get("state"),
+            reqVars.get("organizationName").replace(' ', '-').replace(':', '-'),
+            UUID.randomUUID().toString());
```
i think the thinking is that different organizations *can* in fact have the same name/externalID and the nonce serves to allow them to both exist. what behavior are you hoping for?

jeskew-gov replied:
If someone hits the submit button twice by mistake, I think the only uniqueness constraint they would hit is for user creation (where the email address must be unique).

It would be a slightly better experience if a double submission failed at this step, as that would result in duplicate emails sent and an orphaned organization resource in Okta. I'm not sure if there's anything in the organization resource that should be unique (maybe state + name + CLIA number?), so that's maybe something to ask product about and leave for now as a future enhancement.

jeremyzitomer-usds commented on backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java:
```diff
@@ -47,20 +82,119 @@ public void submitWaitlistRequest(@Valid @RequestBody WaitlistRequest body) thro
+            reqVars.get("zip"),
+            _avs.FACILITY_DISPLAY_NAME);
+    StreetAddress providerAddress =
+        new StreetAddress(
```
cc'ing @TimBest who i suspect knows more than i do on this one -- i know that we collect county for `providerAddress` while we don't for `facilityAddress`, is there a reason why? is this why only `facilityAddress` goes through validation?
